[Bug]: AssertionError at kv_cache_utils.py:1042 — dense draft model + hybrid-attention main (DeltaNet+SWA) fails in unify_kv_cache_spec_page_size

## Summary

Pairing a **dense draft model** (`LocoOperator/LocoOperator-4B`) with a **hybrid-attention main model** (`Qwen/Qwen3-Coder-Next-80B-A3B`, DeltaNet + sliding-window attention) fails at engine init with a bare `AssertionError` in `unify_kv_cache_spec_page_size` — the dense draft's `page_size_bytes` doesn't equal the hybrid main's `max_page_size`.

## Stack trace (live, deterministic across 3 boot attempts)

```
File "/usr/local/lib/python3.12/site-packages/vllm/v1/worker/gpu_model_runner.py", line 5951, in profile_cudagraph_memory
    self._init_minimal_kv_cache_for_profiling()
File "/usr/local/lib/python3.12/site-packages/vllm/v1/worker/gpu_model_runner.py", line 5870, in _init_minimal_kv_cache_for_profiling
    kv_cache_groups = get_kv_cache_groups(self.vllm_config, kv_cache_spec)
File "/usr/local/lib/python3.12/site-packages/vllm/v1/core/kv_cache_utils.py", line 1654, in get_kv_cache_groups
    kv_cache_spec = unify_kv_cache_spec_page_size(kv_cache_spec)
File "/usr/local/lib/python3.12/site-packages/vllm/v1/core/kv_cache_utils.py", line 1042, in unify_kv_cache_spec_page_size
    assert new_spec.page_size_bytes == max_page_size
AssertionError
```

## Reproduction
- **Main**: `Qwen/Qwen3-Coder-Next-80B-A3B-NVFP4` (DeltaNet + SWA hybrid attention, MoE 3B-active / 80B total)
- **Draft**: `LocoOperator/LocoOperator-4B` (4B dense)
- **Args** added to an otherwise-working spec:
  ```
  --speculative-config '{"model": "/draft", "num_speculative_tokens": 3}'
  ```
- **Hardware**: 1× RTX PRO 6000 Blackwell (SM_120, 96 GiB), TP=1, `--kv-cache-dtype fp8_e4m3`
- **Image**: `vllm/vllm-openai:v0.20.2-cu129-ubuntu2404`

Failure reproduces deterministically — 3 boot attempts ~80s apart, identical traceback, weights load to ~60% before the engine hits the assertion.

## Secondary observation

The assertion at `kv_cache_utils.py:1042` has no error message attached (`assert x == y` with no `, f"..."` clause). Operators get a bare `AssertionError` with no expected-vs-actual values. Even just `assert new_spec.page_size_bytes == max_page_size, f"draft spec page_size_bytes={new_spec.page_size_bytes} != main max_page_size={max_page_size}"` would save significant git-archeology time.

## Known historical working config

A fork tracking commit `a29a754d1b` (early 2026) had this exact pairing serving correctly and producing ~2× decode TPS lift on the same hardware class. The regression appears to be downstream of that point — the validation was either added or tightened.

## Ask

1. Is dense-draft + hybrid-attention-main a supported pairing? It worked in the past.
2. If yes — what config are we missing?
3. If no — is there a tracked RFC for adding support, and can we contribute a minimal repro?

## Related
- RFC [#42082](https://github.com/vllm-project/vllm/issues/42082) — Standardize KV-cache Layouts
- Closed RFC [#31634](https://github.com/vllm-project/vllm/issues/31634) — Decouple `page_size_bytes` calculation in AttentionSpec for TPU/RPA

Reproduction host is a Blackwell fleet running `Qwen3-Coder-Next-80B-A3B` as the `AGENTIC_CODE` primary — restoring this spec-decode path would meaningfully reduce per-user decode latency.

<!-- Comment to post on: https://github.com/vllm-project/vllm/issues/43626 -->

Cross-rig reproduction of the page-size assert and a downstream secondary
assert that the existing fix candidates (#37429 / #35062) appear not to
cover. Sharing it here in case it's useful for the eventual fix shape.

## Reproduction on AMD Strix Halo (gfx1151)

Same `AssertionError` at the same line, different stack:

- **Hardware**: AMD Strix Halo, gfx1151 (RDNA 3.5), 40 CUs, UMA
- **Main model**: `cyankiwi/Qwen3.6-27B-AWQ-INT4` (AWQ-INT4 W4A16,
  Qwen3.6-27B base, hybrid DeltaNet GDN + standard attention)
- **Draft model**: `z-lab/Qwen3.6-27B-DFlash` (~2B BF16, interleaved
  SWA+full attention)
- **Trigger**: `--kv-cache-dtype fp8` (with the otherwise-working
  `--speculative-config '{"method":"dflash","num_speculative_tokens":4}'`)
- **vLLM version**: v0.20.0 (bumping to v0.20.2 + v0.21.0 + v0.22.0 all
  reproduce per the same assertion path)

Stack trace path is identical to the original report:
`profile_cudagraph_memory` → `_init_minimal_kv_cache_for_profiling` →
`get_kv_cache_groups` → `unify_kv_cache_spec_page_size` → assert at
`kv_cache_utils.py:1042` (Qwen3.6 lands in the MambaSpec `block_size`-
scaling branch `Sunt-ing` flagged above).

## Local port of #42102 — works, with a 1-line follow-up

Independent of the dense-draft + Qwen3-Next case in this issue, we ported
#42102 (DFlash KV partitioning) onto v0.20.0 to unblock our boot. It
works: engine boots cleanly, KV pool sizes correctly, decode runs.

But the partition alone is not sufficient on our config. With FP8 target +
BF16 drafter we hit a secondary assert downstream of the unify path:

```
File ".../vllm/v1/worker/utils.py", line 157, in __init__   # KVBlockZeroer
    assert page_size_el == cur_page_el, (
AssertionError: Non-uniform page sizes: 827392 vs 1654784
```

The cause: `KVBlockZeroer.__init__` iterates `attn_groups_iter` and
accumulates a *single* `page_size_el` across **all** AttentionSpec
groups before any partition-aware split. Even after a clean
#42102-style partition gives drafter and target separate KV cache
groups, they still pass through the same `KVBlockZeroer` and trip
the assert because BF16 vs FP8 element sizes give pages that differ
2:1.

Workaround that worked end-to-end for us (mirrors the encoder-only
escape hatch already in the runner):

```python
# in GPUModelRunner._init_kv_zero_meta, before the KVBlockZeroer call
from vllm.v1.core.kv_cache_utils import _get_dflash_isolated_layer_names
drafter_layers = _get_dflash_isolated_layer_names(
    self.vllm_config,
    self.compilation_config.static_forward_context.keys(),
)
self.runner_only_attn_layers |= drafter_layers
```

Drafter blocks come zeroed from the CuMem pool allocator and the
per-position seq mask prevents reads of unwritten positions, so
skipping explicit zero-fills for drafter layers is benign.

Whether the right shape upstream is "extend `KVBlockZeroer` to maintain
per-group `page_size_el`" (more correct, more surface) or "skip
isolated drafter layers via `runner_only_attn_layers`" (one-liner,
relies on `_get_dflash_isolated_layer_names` from a partition-style PR
being present) is a design call for the page-size fix author. Pointing
at it here so it's not a surprise after #37429 lands.

## End-to-end benchmark (FP8 KV vs FP16 KV, post-port)

For the negative-result data point — once we got FP8 KV booting on
DFlash + DeltaNet via the port, the throughput was approximately break-
even, not the predicted 30-60%. Three reasons:

1. e4m3 dequant on gfx1151 is software (no hardware FP8 ALU); per K/V
   tile read does `load → cast → multiply scale` in Triton.
2. Drafter still uses BF16 KV (per #42102's drafter dtype override), so
   total KV bandwidth doesn't halve.
3. DFlash per-position acceptance drops ~0.3-0.5 (from ~2.3-2.5 to
   ~1.5-2.1) — quantized target K/V perturbs hidden states enough to
   reduce drafter alignment, partially cancelling bandwidth savings.

Numbers (Qwen3.6-27B AWQ4 + DFlash N=4, `max_num_seqs=1`,
`max_model_len=65536`):

| ctx | FP16 KV (baseline) | FP8 KV (post-port) | delta |
|----:|-------------------:|-------------------:|------:|
|   0 |             14.4 t/s |          14.5 t/s |  +1.0% |
|  8K |             12.8 t/s |          11.8 t/s |  -7.7% |
| 16K |             10.4 t/s |          10.7 t/s |  +3.0% |
| 32K |              7.4 t/s |           7.4 t/s |  +1.0% |

FP8 KV memory savings remain real (~25 GiB at our shape) — the
throughput-neutral cell at 32K shows the fix unlocks the configuration
even if perf isn't always the headline gain. We expect a CUDA host with
hardware FP8 dequant + a higher-acceptance drafter to land squarely in
positive-delta territory.

## On the bare AssertionError

Seconding the request in the original report: the v0.20.0 assert at
line 1042 has no message attached. Once whoever picks this up rebases,
threading the expected-vs-actual page sizes through would save real
operator time:

```python
assert new_spec.page_size_bytes == max_page_size, (
    f"unify could not pad {type(new_spec).__name__} "
    f"to max_page_size={max_page_size} "
    f"(got page_size_bytes={new_spec.page_size_bytes}); "
    f"likely a MambaSpec / AttentionSpec mismatch where block_size "
    f"scaling does not change MambaSpec.page_size_bytes"
)
```

Happy to provide more details on our config / port if useful. The
broader port (the three-layer #42102 fix + the `KVBlockZeroer`
workaround) lives in
[a downstream project](https://github.com/anjoj0/vllm-awq4-qwen) — let
me know if pulling the relevant parts into a focused upstream PR would
help.

--- a/tests/v1/core/test_kv_cache_utils.py
+++ b/tests/v1/core/test_kv_cache_utils.py
@@ -2484,6 +2484,91 @@ def test_unify_hybrid_kv_cache_specs():
         kv_cache_utils.unify_hybrid_kv_cache_specs(kv_cache_spec)
 
 
+def test_unify_kv_cache_spec_page_size_mamba():
+    """Regression test for https://github.com/vllm-project/vllm/issues/43626.
+
+    MambaSpec's page_size_bytes is determined by its state shapes and does not
+    change with block_size, so unify_kv_cache_spec_page_size must pad the Mamba
+    page instead of scaling its block_size. This situation arises when a layer
+    with a page larger than the (already platform-aligned) Mamba page joins the
+    specs, e.g. a dense draft model with more KV heads than the hybrid main
+    model.
+    """
+    # 1. Hybrid main model (Mamba + full attention, pages already aligned at
+    # 16KB) plus a dense draft model layer with a 2x larger page (32KB).
+    # Reproduces the bare AssertionError from #43626: 32768 % 16384 == 0, so
+    # the old code scaled the Mamba block_size, which left page_size_bytes
+    # unchanged at 16384.
+    mamba_spec = new_mamba_spec()  # page_size_bytes = 16384
+    main_attn_spec = new_kv_cache_spec()  # page_size_bytes = 16384
+    draft_attn_spec = new_kv_cache_spec(num_kv_heads=4)  # page_size_bytes = 32768
+    assert mamba_spec.page_size_bytes == main_attn_spec.page_size_bytes == 16384
+    assert draft_attn_spec.page_size_bytes == 32768
+
+    unified = kv_cache_utils.unify_kv_cache_spec_page_size(
+        {
+            "mamba_layer": mamba_spec,
+            "main_attn_layer": main_attn_spec,
+            "draft_attn_layer": draft_attn_spec,
+        }
+    )
+    # Mamba page is padded; block_size (caching granularity) is unchanged.
+    assert unified["mamba_layer"].page_size_bytes == 32768
+    assert unified["mamba_layer"].page_size_padded == 32768
+    assert unified["mamba_layer"].block_size == mamba_spec.block_size
+    # Attention layer with smaller page still unifies by scaling block_size.
+    assert unified["main_attn_layer"].page_size_bytes == 32768
+    assert unified["main_attn_layer"].block_size == 2 * main_attn_spec.block_size
+    assert unified["main_attn_layer"].page_size_padded is None
+    # Layer already at max page size is unchanged.
+    assert unified["draft_attn_layer"] == draft_attn_spec
+
+    # 2. Mamba page already padded by the platform (state smaller than the
+    # padded page); the padding is re-applied at the new maximum.
+    padded_mamba_spec = new_mamba_spec(
+        shapes=((2, 256), (3, 32, 32)), page_size_padded=16384
+    )
+    assert padded_mamba_spec.page_size_bytes == 16384
+    unified = kv_cache_utils.unify_kv_cache_spec_page_size(
+        {
+            "mamba_layer": padded_mamba_spec,
+            "draft_attn_layer": draft_attn_spec,
+        }
+    )
+    assert unified["mamba_layer"].page_size_bytes == 32768
+    assert unified["mamba_layer"].page_size_padded == 32768
+
+    # 3. Mamba page that does not evenly divide the maximum page size is
+    # padded as well (the divisibility constraint only applies to block_size
+    # scaling).
+    odd_mamba_spec = new_mamba_spec(shapes=((6144,),))
+    assert odd_mamba_spec.page_size_bytes == 24576
+    assert 32768 % odd_mamba_spec.page_size_bytes != 0
+    unified = kv_cache_utils.unify_kv_cache_spec_page_size(
+        {
+            "mamba_layer": odd_mamba_spec,
+            "draft_attn_layer": draft_attn_spec,
+        }
+    )
+    assert unified["mamba_layer"].page_size_bytes == 32768
+
+    # 4. Attention layers with non-divisible page sizes still raise.
+    with pytest.raises(NotImplementedError):
+        kv_cache_utils.unify_kv_cache_spec_page_size(
+            {
+                "attn_layer": new_kv_cache_spec(block_size=24),  # 24576
+                "draft_attn_layer": draft_attn_spec,  # 32768
+            }
+        )
+
+    # 5. Uniform page sizes are returned unchanged.
+    specs = {
+        "mamba_layer": new_mamba_spec(),
+        "attn_layer": new_kv_cache_spec(),
+    }
+    assert kv_cache_utils.unify_kv_cache_spec_page_size(specs) == specs
+
+
 def test_hma_not_disabled_when_kv_events_enabled():
     """
     Test enabling KV events must not force disable_hybrid_kv_cache_manager to True.

--- a/vllm/v1/core/kv_cache_utils.py
+++ b/vllm/v1/core/kv_cache_utils.py
@@ -1051,11 +1051,12 @@ def unify_kv_cache_spec_page_size(
 ) -> dict[str, KVCacheSpec]:
     """
     Unify the page size of the given KVCacheSpec. If the page size of all layers
-    are the same, return the original KVCacheSpec. If not same, first try to
-    unify page size by increasing the block size of layers with smaller page
-    size. If a smaller attention page does not evenly divide the maximum page
-    size, keep its logical block size and pad its physical page instead --- but
-    only for attention layers whose backend opts in via
+    are the same, return the original KVCacheSpec. If not same, unify the page
+    size by increasing the block size of layers with smaller page size. Two
+    cases cannot be unified by block size alone and pad their physical page to
+    the maximum instead: Mamba layers, whose page size comes from state shapes
+    and is independent of block size; and attention layers whose page does not
+    evenly divide the maximum and whose backend opts in via
     ``AttentionSpec.indexes_kv_by_block_stride`` (the padded page is read through
     a strided view, which not every backend handles). Raise NotImplementedError
     if failed to unify the page size.
@@ -1076,6 +1077,16 @@ def unify_kv_cache_spec_page_size(
     for layer_name, layer_spec in kv_cache_spec.items():
         if layer_spec.page_size_bytes == max_page_size:
             new_kv_cache_spec[layer_name] = layer_spec
+        elif isinstance(layer_spec, MambaSpec):
+            # MambaSpec's page size is determined by its state shapes and does
+            # not scale with block_size, so pad the page instead. This is the
+            # same padding mechanism the platform uses to align Mamba pages
+            # with the main model's attention page size; it is needed here
+            # when another layer (e.g. from a draft model) has a larger page
+            # than the already-aligned Mamba page.
+            new_spec: KVCacheSpec = replace(layer_spec, page_size_padded=max_page_size)
+            assert new_spec.page_size_bytes == max_page_size
+            new_kv_cache_spec[layer_name] = new_spec
         else:
             layer_page_size = layer_spec.page_size_bytes
             if max_page_size % layer_page_size == 0:

